api/layer/object.go

```diff
@@ -71,70 +66,101 @@ type (
+
+	res, err := n.neoFS.SelectObjects(ctx, prm)
 
-	ids, err := n.pool.SearchObject(ctx, searchParams, n.CallOptions(ctx)...)
-	return ids, n.transformNeofsError(ctx, err)
+	return res, n.transformNeofsError(ctx, err)
 }
 
-func newAddress(cid *cid.ID, oid *object.ID) *object.Address {
-	address := object.NewAddress()
-	address.SetContainerID(cid)
-	address.SetObjectID(oid)
-	return address
+func newAddress(cid *cid.ID, oid *oid.ID) *address.Address {
+	addr := address.NewAddress()
+	addr.SetContainerID(cid)
+	addr.SetObjectID(oid)
+	return addr
 }
 
 // objectHead returns all object's headers.
-func (n *layer) objectHead(ctx context.Context, cid *cid.ID, oid *object.ID) (*object.Object, error) {
-	ops := new(client.ObjectHeaderParams).WithAddress(newAddress(cid, oid)).WithAllFields()
-	obj, err := n.pool.GetObjectHeader(ctx, ops, n.CallOptions(ctx)...)
-	return obj, n.transformNeofsError(ctx, err)
+func (n *layer) objectHead(ctx context.Context, idCnr *cid.ID, idObj *oid.ID) (*object.Object, error) {
+	prm := PrmObjectRead{
+		Container:  *idCnr,
+		Object:     *idObj,
+		WithHeader: true,
+	}
+
+	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+
+	res, err := n.neoFS.ReadObject(ctx, prm)
+	if err != nil {
+		return nil, n.transformNeofsError(ctx, err)
+	}
+
+	return res.Head, nil
 }
 
-// objectGetWithPayloadWriter and write it into provided io.Reader.
-func (n *layer) objectGetWithPayloadWriter(ctx context.Context, p *getParams) (*object.Object, error) {
-	// prepare length/offset writer
-	w := newWriter(p.Writer, p.offset, p.length)
-	ops := new(client.GetObjectParams).WithAddress(newAddress(p.cid, p.oid)).WithPayloadWriter(w)
-	obj, err := n.pool.GetObject(ctx, ops, n.CallOptions(ctx)...)
-	return obj, n.transformNeofsError(ctx, err)
+// writes payload part of the NeoFS object to the provided io.Writer.
+// Zero range corresponds to full payload (panics if only offset is set).
+func (n *layer) objectWritePayload(ctx context.Context, p getParams) error {
+	prm := PrmObjectRead{
+		Container:    *p.cid,
+		Object:       *p.oid,
+		WithPayload:  true,
+		PayloadRange: [2]uint64{p.off, p.ln},
+	}
+
+	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+
+	res, err := n.neoFS.ReadObject(ctx, prm)
+	if err == nil {
+		defer res.Payload.Close()
+
+		if p.ln == 0 {
+			p.ln = 4096 // configure?
+		}
+
+		// alloc buffer for copying
+		buf := make([]byte, p.ln) // sync-pool it?
+
+		// copy full payload
+		_, err = io.CopyBuffer(p.w, res.Payload, buf)
+	}
+
+	return n.transformNeofsError(ctx, err)
 }
 
 // objectGet returns an object with payload in the object.
-func (n *layer) objectGet(ctx context.Context, cid *cid.ID, oid *object.ID) (*object.Object, error) {
-	ops := new(client.GetObjectParams).WithAddress(newAddress(cid, oid))
-	obj, err := n.pool.GetObject(ctx, ops, n.CallOptions(ctx)...)
-	return obj, n.transformNeofsError(ctx, err)
-}
+func (n *layer) objectGet(ctx context.Context, addr *address.Address) (*object.Object, error) {
```

Comment:
It's expected this method to return `object.Object` with not empty payload field but as I can see we return only headers

Reply:
`NeoFS.ReadObject` is used correctly, but implementation doesn't follow the requirement. Fixed.